### ISSUE TITLE
Using mongoose model discriminator

### ISSUE THREAD

**Issue by abouroubi** [szokodiakos/typegoose#294]
One of the best practices in the NoSQL databases, is to use same collection for multiple related models.

Mongoose provide a function to to that, ```model.discriminator()```, the doc can be found here : https://mongoosejs.com/docs/discriminators.html

This function is very convient, because it adds a discriminator key to the schema, and automatically filters on this key when using find, update or other.

Is it possible to implement this in typegoose ?

I tried using this function from my base model, but it takes the schema of the child model, and unfortunately typegoose doesn't expose the getSchema function it's private, I wonder why this choice was made.

Here is my example :
```Typescript
 const adminSchema = new Admin().getModelForClass(Admin, {
      schemaOptions: { ...schemaOptions, discriminatorKey: 'role' },
    }).schema;

    const adminModel = BaseUserModel.discriminator(
      Admin.name,
      adminSchema,
    );

```

This example obviously don't work for multiple reasons :
1) getModelForClass compiles the AdminModel and discriminator function tries to di the same, mongoose throws an exception because of that.

2) The discriminator function returns a mongoose model document, which is not a Typegosse document.

**Comment by hasezoey** [comment]
i'm new to the `model.discriminator()` thing, but i think i found what you search for
( you can skip down to the "import part")

```ts
import * as mongoose from "mongoose";
import { prop, Typegoose } from "typegoose";

class DummyType1 extends Typegoose {
    @prop({ required: true, default: "prop1_default" })
    public prop1!: string;
}

class DummyType2 extends DummyType1 {
    @prop({ required: true })
    public prop2!: string;
}

class DummyType3 extends DummyType1 {
    @prop({ required: true })
    public prop3!: string;
}

//  here the important part starts
const DummyModel1 = new DummyType1().getModelForClass(DummyType1);
const DummyModel2 = DummyModel1.discriminator("D2", new DummyType2().getModelForClass(DummyType2).schema);
const DummyModel3 = DummyModel1.discriminator("D3", new DummyType3().getModelForClass(DummyType3).schema);
//  here it ends

(async () => {
    mongoose.set("debug", true);
    await mongoose.connect(`mongodb://mongodb:27017/`, {
        useNewUrlParser: true,
        useFindAndModify: true,
        useCreateIndex: true,
        dbName: "verify294",
        user: "user",
        pass: "passwd",
        authSource: "admin",
        autoIndex: true
    });

    await new DummyModel2({ prop2: "prop2" } as DummyType2).save();
    await new DummyModel3({ prop3: "prop3" } as DummyType3).save();

    await mongoose.disconnect();
})();

```

**Comment by abouroubi** [comment]
The problem with that implementation is that you're creating two model for the same class.

Calling `.getModelForClass(DummyType2).schema)` will create a model inside mongoose called "DummyType2" based on schema of "DummyType2" and calling `DummyModel1.discriminator("D2", new DummyType2()` will create the same model just different name, here it will be "D2".

I'm sure there is a better way to solve this issue.

**Comment by hasezoey** [comment]
this can be closed in favor of #332 
@Ben305
